### PR TITLE
bpo-41835: Speedup to dict vectorcall with keyword arguments.

### PR DESCRIPTION
**Pull request by Marco-Sulla**
I added a specialized version of insertdict, insertdict_init, that assumes any needed resize to the dict is done before the inserting.

Also done some minor and cosmetic change.

https://bugs.python.org/issue41835

### CODE REVIEW

**Review by eamanu — CHANGES_REQUESTED** [review]
Please, this PR definitely need a bpo for discussion, and NEWS.

**Comment by sweeneyde** [comment]
Do you have a microbenchmark for what specific thing is being optimized? I'd recommend using [pyperf](https://pypi.org/project/pyperf/) and posting the results to an issue at bugs.python.org.

From the [Dev Guide](https://devguide.python.org/pullrequest/#making-good-commits):

* Do not do cosmetic changes to unrelated code in the same commit as some feature/bugfix.

It seems like most of the diff is filled with these. I think the general guideline is to keep the diff as small as possible. I would think that the PR should limit changing cosmetics (`const` qualifiers, delayed declarations, whitespace, braces, changing `if`s to `assert`s) to the couple of functions that the PR is to focus on. Finally, the change to .gitignore looks unrelated to the suggestion.

Is there anything positive accomplished by moving behavior from public functions like `PyDict_SetItem` and `PyDict_GetItemWithError` into new identical static-scope functions? If anything, I think the existing usage of the public functions is good because it gives examples of how the API is to be used outside of this file.

File: .gitignore [hunk 1]
```diff
@@ -21,6 +21,7 @@
 *.profclang?
 *.profraw
 *.dyn
+*.ipynb
```
Comment (methane) [inline .gitignore]: Please move this to your `core.excludesfile` file.

Comment (Marco-Sulla) [reply .gitignore]: OK, I removed it.

**Comment by Marco-Sulla** [comment]
@sweeneyde: I removed all the minor and cosmetic changes.

About PyDict_SetItem and the other one, I followed the example of setobject.c, where all the public functions calls the static internal functions. IMHO it's more clear. Furthermore, should it save you some nanoseconds (static objects are allocated in the stack)?

**Comment by methane** [comment]
@Marco-Sulla Please create bpo issue and post benchmark&result to it.

**Comment by Marco-Sulla** [comment]
I already created it: https://bugs.python.org/issue41835

File: Objects/dictobject.c [hunk 2]
```diff
@@ -3446,21 +3543,44 @@ dict_vectorcall(PyObject *type, PyObject * const*args,
...
         }
+        
+        empty = 0;
         args++;
     }
+    
+    Py_ssize_t kw_size = (kwnames == NULL
+        ? 0
+        : PyTuple_GET_SIZE(kwnames)
+    );
```
Comment (methane) [inline Objects/dictobject.c]: Move this into `if (kwnames != NULL) {`

**Review by methane — COMMENTED** [review]
Remove trailing whitespaces.

I am not sure this is the best performance/complexity way. So I am trying some other ideas. Please wait.

File: Objects/dictobject.c (hunk 2 above)
Comment (Marco-Sulla) [reply Objects/dictobject.c]: Done. I also removed the empty lines :)

**Pull request closed without merging**